### ISSUE TITLE
Cannot pull port (requestParsingIn) twice

### ISSUE THREAD

**Issue by sambogoat** [akka/akka#21008]
Environment: Java 8, scala 2.11.7, akka 2.4.8

Testing with:
curl -v -F"a=@test1.dat" -F"b=@test2.dat" http://localhost:7777/test

The following code results in a Timeout Error:

```
object AkkaTest extends App {
  implicit val system = ActorSystem()
  implicit val materializer = ActorMaterializer()
  implicit val executor = system.dispatcher

  val route = encodeResponse {
    (post & path("test") & entity(as[Multipart.FormData])) { entity =>
      complete {
        val sources = entity.parts.map(e => e.entity.dataBytes)
        val f = sources.runForeach { source =>
          val inputStream = source.runWith(StreamConverters.asInputStream())
          try {
            val tmp = Files.createTempFile("a", "b")
            Files.copy(inputStream, tmp, StandardCopyOption.REPLACE_EXISTING)
            Files.delete(tmp)
          } finally {
            inputStream.close()
          }
        }
        f.map(_ => "ok")
      }
    }
  }
  import akka.http.scaladsl.Http
  Http().bindAndHandle(route, "0.0.0.0", 7777)
```

Whereas, this works as expected (using runFold and ByteArrrayInputStream):

```
object AkkaTest extends App {
  implicit val system = ActorSystem()
  implicit val materializer = ActorMaterializer()
  implicit val executor = system.dispatcher

  val route = encodeResponse {
    (post & path("test") & entity(as[Multipart.FormData])) { entity =>
      complete {
        val sources = entity.parts.map(e => e.entity.dataBytes)
        val f = sources.runForeach { source =>
          val tmp = Files.createTempFile("a", "b")
          source.runFold(ByteString.empty)((bs, c) => bs ++ c)
            .foreach { bs =>
              val inputStream = new ByteArrayInputStream(bs.toArray)
              try {
                Files.copy(inputStream, tmp, StandardCopyOption.REPLACE_EXISTING)
                Files.delete(tmp)
              } finally {
                inputStream.close()
              }
            }
        }
        f.map(_ => "ok")
      }
    }
  }
  import akka.http.scaladsl.Http
  Http().bindAndHandle(route, "0.0.0.0", 7777)
}
```

Exception:
```
java.io.IOException: Timeout on waiting for new data
	at akka.stream.impl.io.InputStreamAdapter$$anonfun$read$1.apply$mcI$sp(InputStreamSinkStage.scala:144) ~[akka-stream_2.11-2.4.8.jar:na]
	at akka.stream.impl.io.InputStreamAdapter$$anonfun$read$1.apply(InputStreamSinkStage.scala:129) ~[akka-stream_2.11-2.4.8.jar:na]
	at akka.stream.impl.io.InputStreamAdapter$$anonfun$read$1.apply(InputStreamSinkStage.scala:129) ~[akka-stream_2.11-2.4.8.jar:na]
	at akka.stream.impl.io.InputStreamAdapter.executeIfNotClosed(InputStreamSinkStage.scala:109) ~[akka-stream_2.11-2.4.8.jar:na]
	at akka.stream.impl.io.InputStreamAdapter.read(InputStreamSinkStage.scala:129) ~[akka-stream_2.11-2.4.8.jar:na]
	at akka.stream.impl.io.InputStreamAdapter.read(InputStreamSinkStage.scala:120) ~[akka-stream_2.11-2.4.8.jar:na]
	at java.nio.file.Files.copy(Files.java:2908) ~[na:1.8.0_45]
	at java.nio.file.Files.copy(Files.java:3027) ~[na:1.8.0_45]
	at com.scb.sabre.store.http.AkkaTest$$anonfun$15$$anonfun$apply$69$$anonfun$apply$70$$anonfun$17.apply(
....
2016-07-20 10:02:13,202 - [ERROR] - [akka.actor.RepointableActorRef] [undefined] [default-akka.actor.default-dispatcher-2] - Error in stage [akka.http.impl.engine.server.HttpServerBluePrint$ControllerStage@5455fb41]: requirement failed: Cannot pull port (requestParsingIn) twice
java.lang.IllegalArgumentException: requirement failed: Cannot pull port (requestParsingIn) twice
	at scala.Predef$.require(Predef.scala:219) ~[scala-library-2.11.6.jar:na]
	at akka.stream.stage.GraphStageLogic.pull(GraphStage.scala:348) ~[akka-stream_2.11-2.4.8.jar:na]
	at akka.http.impl.engine.server.HttpServerBluePrint$ControllerStage$$anon$12$$anon$15.onPush(HttpServerBluePrint.scala:428) ~[akka-http-core_2.11-2.4.8.jar:na]
	at akka.stream.impl.fusing.GraphInterpreter.processElement$1(GraphInterpreter.scala:590) ~[akka-stream_2.11-2.4.8.jar:na]
	at akka.stream.impl.fusing.GraphInterpreter.processEvent(GraphInterpreter.scala:601) ~[akka-stream_2.11-2.4.8.jar:na]
	at akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:542) ~[akka-stream_2.11-2.4.8.jar:na]
	at akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:471) ~[akka-stream_2.11-2.4.8.jar:na]

```

Possibly related to:
https://github.com/akka/akka/issues/19392
https://github.com/akka/akka/issues/20642

**Comment by drewhk** [comment]
The "cannot pull port" part is a bug on our side. As a sidenote, I don't find it a very good idea to convert streams to InputStream as it will involve blocking and might actually deadlock your server in certain unfortunate cases. Why not just use the file sink to stream directly the data to the file you want? It is much safer and likely faster.

**Comment by sambogoat** [comment]
I actually chose writing to File just to exemplify the problem in this ticket. In the real scenario we call a 3rd party library that requires an InputStream and we use StreamConverters for this.

**Comment by johanandren** [comment]
In general replying before you have consumed all (or possibly any) request entity bytes like you do in the failing sample is unsafe, it usually leads to a log entry like this, which is not displayed in this case for some reason:

```
Sending an 2xx 'early' response before end of request was received...
Note that the connection will be closed after this response. Also, many clients will not read early responses! 
Consider only issuing this response after the request data has been completely read!
```

Which means that the second solution where you do not reply to the client until you have consumed each multipart.

I'm looking into what is special with the multipart and this special case, but even if it gets fixed you should go for the second option.

**Comment by johanandren** [comment]
Actual minimal repeater, without any blocking stuff:

```
 val route =
    extractRequest { req =>
      complete {
        req.entity.dataBytes.runForeach { chunk =>
          // start consuming multiparts, but fail
          throw new RuntimeException("argh")
        }.map[HttpResponse](_ => ???)
      }
    }
  Http().bindAndHandle(route, "0.0.0.0", 7777)
```

When run with for example:
```
curl -F "a=@README.md" -F "b=@CONTRIBUTING.md" http://localhost:7777/test
```
this triggers the `Controller` stage to pull/push when it should't. The exact same but as a partial function combined with `bindAndHandleAsync` does not trigger the error, so there is something special in the implicit conversion causing it (which is weird because that essentially just makes a `HttpRequest => Future[HttpResponse]` out of the route).

**Comment by johanandren** [comment]
Even more minimal repeater:
```
 Http().bindAndHandleAsync({ req =>
    req.entity.dataBytes.runForeach { chunk =>
      // start consuming entity, but fail on first chunk
      throw new RuntimeException("argh")
    }.map(_ => HttpResponse()).recover {
      case ex => HttpResponse(status = StatusCodes.InternalServerError)
    }
  }, "0.0.0.0", 7777)
```

**Comment by tummybunny** [comment]
```scala
val route = encodeResponse {
    (post & path("test") & entity(as[Multipart.FormData])) { entity =>
      complete {
        val sources = entity.parts.map(e => e.entity.dataBytes)
        val f = sources.runForeach { source =>
          val inputStream = source.runWith(StreamConverters.asInputStream())
          try {
            val tmp = Files.createTempFile("a", "b")
            Files.copy(inputStream, tmp, StandardCopyOption.REPLACE_EXISTING)
            Files.delete(tmp)
          } finally {
            inputStream.close()
          }
        }
        f.map(_ => "ok")
      }
    }
  }
```

This doesn't seem to be fixed in 2.4.9 because it is still timing out.
```
java.io.IOException: Timeout on waiting for new data
	at akka.stream.impl.io.InputStreamAdapter$$anonfun$read$1.apply$mcI$sp(InputStreamSinkStage.scala:144) ~[akka-stream_2.11-2.4.9.jar:na]
	at akka.stream.impl.io.InputStreamAdapter$$anonfun$read$1.apply(InputStreamSinkStage.scala:129) ~[akka-stream_2.11-2.4.9.jar:na]
	at akka.stream.impl.io.InputStreamAdapter$$anonfun$read$1.apply(InputStreamSinkStage.scala:129) ~[akka-stream_2.11-2.4.9.jar:na]
	at akka.stream.impl.io.InputStreamAdapter.executeIfNotClosed(InputStreamSinkStage.scala:109) ~[akka-stream_2.11-2.4.9.jar:na]
	at akka.stream.impl.io.InputStreamAdapter.read(InputStreamSinkStage.scala:129) ~[akka-stream_2.11-2.4.9.jar:na]
	at akka.stream.impl.io.InputStreamAdapter.read(InputStreamSinkStage.scala:120) ~[akka-stream_2.11-2.4.9.jar:na]
	at java.nio.file.Files.copy(Files.java:2908) ~[na:1.8.0_45]
	at java.nio.file.Files.copy(Files.java:3027) ~[na:1.8.0_45]
	at xyz.AkkaTest$$anonfun$6.apply(AkkaTest.scala:135) ~[classes/:na]
	at xyz.AkkaTest$$anonfun$6.apply(AkkaTest.scala:131) ~[classes/:na]
	at akka.stream.impl.fusing.Map$$anon$5.onPush(Ops.scala:42) ~[akka-stream_2.11-2.4.9.jar:na]
	at akka.stream.impl.fusing.GraphInterpreter.processPush(GraphInterpreter.scala:755) ~[akka-stream_2.11-2.4.9.jar:na]
	at akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:657) ~[akka-stream_2.11-2.4.9.jar:na]
	at akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:470) ~[akka-stream_2.11-2.4.9.jar:na]
	at akka.stream.impl.fusing.GraphInterpreterShell.receive(ActorGraphInterpreter.scala:422) ~[akka-stream_2.11-2.4.9.jar:na]
	at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:602) ~[akka-stream_2.11-2.4.9.jar:na]
	at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$shortCircuitBatch(ActorGraphInterpreter.scala:593) ~[akka-stream_2.11-2.4.9.jar:na]
	at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:618) ~[akka-stream_2.11-2.4.9.jar:na]
	at akka.actor.Actor$class.aroundReceive(Actor.scala:484) ~[akka-actor_2.11-2.4.9.jar:na]
	at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:528) ~[akka-stream_2.11-2.4.9.jar:na]
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526) [akka-actor_2.11-2.4.9.jar:na]
	at akka.actor.ActorCell.invoke(ActorCell.scala:495) [akka-actor_2.11-2.4.9.jar:na]
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257) [akka-actor_2.11-2.4.9.jar:na]
	at akka.dispatch.Mailbox.run(Mailbox.scala:224) [akka-actor_2.11-2.4.9.jar:na]
	at akka.dispatch.Mailbox.exec(Mailbox.scala:234) [akka-actor_2.11-2.4.9.jar:na]
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260) [na:na]
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339) [na:na]
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979) [na:na]
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107) [na:na]
```

**Comment by johanandren** [comment]
@tummybunny please open a new ticket instead of commenting on this closed one.
